Encode filter parameter as a base64-encoded JSON string in List requests

neuromage opened the pull request:
Generate new swagger definitions and use these to generate the frontend
APIs using `npm run apis`.

This is to support filtering in List requests, as the current
grpc-gateway swagger generator tool does not support repeated fields in
requests used in GET endpoints.

Using POST lets us get around the problem in issue grpc-ecosystem/grpc-gateway#492

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/kubeflow/pipelines/563)
<!-- Reviewable:end -->

neuromage commented:
/assign @yebrahim 
/assign @rileyjbauer

yebrahim commented:
The frontend needs to change the API calls to pass the object. I can help take a look at this soon.

yebrahim commented on frontend/src/apis/experiment/api.ts:
```diff
@@ -191,45 +357,6 @@ export interface ProtobufEmpty {
-         * @param {*} [options] Override http request option.
-         * @throws {RequiredError}
-         */
-        createExperiment(body: ApiExperiment, options: any = {}): FetchArgs {
```
Why are these removed?

neuromage replied:
Thanks for the heads up. This was indeed messed up, because both create and list used the same endpoints with differing verbs (POST and GET).

I decided to go another route: encode the filter as a simple JSON string and put it in a single parameter. Maybe this was what you were alluding to earlier, and I think this should work better, as it doesn't break any of the existing semantics. See what you think.

yebrahim commented on backend/src/apiserver/server/experiment_server.go:
```diff
@@ -44,7 +46,23 @@ func (s *ExperimentServer) ListExperiment(ctx context.Context, request *api.List
 		opts, err = list.NewOptionsFromToken(request.PageToken, int(request.PageSize))
 	} else {
-		opts, err = list.NewOptions(&model.Experiment{}, int(request.PageSize), request.SortBy, request.Filter)
+		var f *api.Filter
```
Any chance we can reuse this code instead of copying it?

neuromage replied:
Thanks! I consolidated this logic into a new function, and also added tests for it. PTAL.

neuromage commented on backend/src/apiserver/server/list_request_util_test.go:
```diff
@@ -175,3 +177,40 @@ func TestParseSortByQueryString_StringTooLong(t *testing.T) {
+	}
+
+	got, err := parseAPIFilter(in)
+	if !cmp.Equal(got, want) || err != nil {
```
True, but I actually prefer the cmp package, as it prints a more precise diff. It's also maintained and recommended by the Go team.

rileyjbauer commented on backend/api/go_http_client/experiment_client/experiment_service/list_experiment_parameters.go:
```diff
@@ -126,6 +132,17 @@ func (o *ListExperimentParams) SetHTTPClient(client *http.Client) {
+
+// SetFilter adds the filter to the list experiment params
+func (o *ListExperimentParams) SetFilter(filter *string) {
+	o.Filter = filter
```
If this function is so simple, and not widely used, why not just inline it?

neuromage replied:
This is auto-generated by the swagger tool. Everything under api/go_http_client and api/go_client is autogenerated.

rileyjbauer replied:
Ah, sorry, I didn't realize they were auto-generated

neuromage replied:
Sorry, yeah, it's definitely not obvious :-)

rileyjbauer commented:
/lgtm

rileyjbauer commented:
/test kubeflow-pipeline-e2e-test

IronPan commented on backend/src/apiserver/server/experiment_server.go:
```diff
@@ -44,7 +44,12 @@ func (s *ExperimentServer) ListExperiment(ctx context.Context, request *api.List
 		opts, err = list.NewOptionsFromToken(request.PageToken, int(request.PageSize))
 	} else {
-		opts, err = list.NewOptions(&model.Experiment{}, int(request.PageSize), request.SortBy, request.Filter)
+		f, err := parseAPIFilter(request.Filter)
```
not quite understand the if else statement here. does it means only one of pagetoken and filter can exist?
Are filtered results paginated?

neuromage replied:
pagetoken encapsulates both sorting and filtering, so if you already have a non-empty page token, we ignore further filtering + sorting criteria.

IronPan commented on backend/src/apiserver/server/list_request_util.go:
```diff
@@ -157,3 +158,27 @@ func deserializePageToken(pageToken string) (*common.Token, error) {
+		return nil, nil
+	}
+
+	errorF := func(err error) (*api.Filter, error) {
```
is any good reason of having this function instead of returning errors directly?

neuromage replied:
Avoids duplicate in error handling, since the message is the same. It's a very idiomatic Go thing to do.

IronPan commented on backend/src/apiserver/server/list_request_util.go:
```diff
@@ -157,3 +158,27 @@ func deserializePageToken(pageToken string) (*common.Token, error) {
+	if err := jsonpb.UnmarshalString(string(b), f); err != nil {
+		return errorF(err)
+	}
+	return f, nil
```
can we add additional validation to make sure the filter format is correct?
e.g. value missing, op missing etc.

neuromage replied:
The api filter will get parsed into a filter.Filter object that can actually be used to do filtering. That parsing will sanity check to ensure the filter is well formed and return an error otherwise.

neuromage commented:
Thanks @rileyjbauer for pointing out that the Filter definitions for Swagger went missing with this change. This was caused by the fact that Filter is only implicitly used by the API, so protoc-gen-swagger will not generate Swagger files correctly for this proto. I modified the Makefile to hack in a dummy service during generation to ensure we get the Swagger definitions. And I updated the `npm run apis` script to generate TS interface definitions for filter as well. 

PTAL.

neuromage commented:
/retest

neuromage commented on backend/api/Makefile:
```diff
@@ -52,6 +52,13 @@ all: dependencies
 		--grpc-gateway_out=logtostderr=true:go_client \
 		*.proto
 
+  # Filter.proto is implicitly used by clients and server, and transmitted as a
```
Done.

Ark-kun replied:
> nit: indentation (this file uses tabs for some reason).

It's not a nit. Makefile requires tabs for receipt prefixes and will misbehave if they are not there.

yebrahim replied:
Weird, but good to know.

neuromage commented:
Yeah, it's a pretty ugly hack. 

I did have the dummy service in initially, but it was generating a corresponding client/server code in Go, which didn't seem like the right thing to do. Hence, this hack :-(

neuromage commented:
/retest

rileyjbauer commented:
/test kubeflow-pipeline-e2e-test

yebrahim commented:
How about setting up "go vet -shadow" for catching these errors. I think shadowing is bad practice anyway. 
@neuromage @IronPan

neuromage commented:
Agreed! We can set this up in Travis as part of the presubmit. If you're ok with it, I can try to do that, and also fix any existing shadow errors.

yebrahim commented:
/lgtm

neuromage commented:
yes, definitely, I meant in a separate PR. This is one is good to go @yebrahim @IronPan

IronPan commented:
/lgtm
/approve
assuming that you will have a separate PR to validate the query string as we discussed offline
- If page token doesn't exist, validate the filter and sort by fields.
- If page token exist, either
   - filter and sort by fields exist and they are valid and equals to the one encoded in page token
   - or they can be opt-out and no validation is needed in this case.
